src/util.h

```diff
@@ -341,15 +341,30 @@ class MaybeStackBuffer {
     AllocateSufficientStorage(storage);
   }
 
+  bool IsAllocated() {
+    return buf_ != buf_st_;
+  }
+
   ~MaybeStackBuffer() {
-    if (buf_ != buf_st_)
+    // Note: if Release() is called, this memory will not be freed here.
+    if (!released_ && buf_ != buf_st_)
       free(buf_);
   }
 
+  // Caution: Causes allocations to not be freed when instances
+  // are deconstructed. By calling Release, ownership (and responsibility
+  // for) the allocated internal buffer can be transferred to another
+  // object (like a Buffer) without copying.
+  bool Release() {
+    released_ = IsAllocated();
+    return released_;
+  }
+
  private:
   size_t length_;
   T* buf_;
   T buf_st_[kStackStorageSize];
+  bool released_;
```

Comment:
You could probably just assign `buf_` to `buf_st_` when transferring the memory ownership. If you want to stick with a property, I’d suggest moving it in front of `buf_st_` so that `length_`, `buf_`, `released_` (and in the common case, the actual data, too) fit into a single cache line.